File: BinaryGCD/Python/BinaryGCD.py
def gcd(a, b, res):
	if a == b:
		return res * a
	elif (a % 2 == 0) and (b % 2 == 0):
		return gcd(a // 2, b // 2, 2 * res)
	elif (a % 2 == 0):
		return gcd(a // 2, b, res)
	elif (b % 2 == 0):
		return gcd(a, b // 2, res)
	elif a > b:
		return gcd(a - b, b, res)
	else:
		return gcd(a, b - a, res)

print(gcd(258, 321, 1))

File: algorithms/Go/BinaryGCD/binarygcd.go
package gcd

func Gcd(a, b int) int {
	if a == 0 {
		return b
	}

	if b == 0 {
		return a
	}

	pow2 := uint(0)

	for ; ((a | b) & 1) == 0; pow2++ {
		a >>= 1
		b >>= 1
	}

	for (a & 1) == 0 {
		a >>= 1
	}

	for b != 0 {
		for (b & 1) == 0 {
			b >>= 1
		}

		if a > b {
			a, b = b, a
		}

		b -= a
	}

	return a << pow2
}
